In [2]:
import requests
import json

# Part 1

In [3]:
sample_json = '''
{
  "station": "USC00305800",
  "name": "New York Central Park",
  "location": {
    "latitude": 40.7789,
    "longitude": -73.9692
  },
  "observations": [
    {"date": "2023-01-01", "temperature": 32, "precipitation": 0.0},
    {"date": "2023-01-02", "temperature": 28, "precipitation": 0.5},
    {"date": "2023-01-03", "temperature": 35, "precipitation": 0.0},
    {"date": "2023-01-04", "temperature": 38, "precipitation": 0.2},
    {"date": "2023-01-05", "temperature": 41, "precipitation": 0.0}
  ]
}
'''


In [4]:
data = json.loads(sample_json)

In [5]:
print("Station:", data['station'])

Station: USC00305800


In [6]:
print("Location:", data['location'])

Location: {'latitude': 40.7789, 'longitude': -73.9692}


In [7]:
print("First observation:", data['observations'][0])

First observation: {'date': '2023-01-01', 'temperature': 32, 'precipitation': 0.0}


## Task 1

In [9]:
print("Date, Temperature")
for obs in data["observations"]: print(obs['date'], obs["temperature"])

Date, Temperature
2023-01-01 32
2023-01-02 28
2023-01-03 35
2023-01-04 38
2023-01-05 41


## Task 2

In [11]:
total_temp = 0 
count = 0

In [12]:
for obs in data[ 'observations']:
    total_temp += obs ['temperature']
    count+=1

In [14]:
average_temp = total_temp / count
print(f"The average temperature is: {average_temp}")

The average temperature is: 34.8


## Task 3

In [16]:
print ("\nDays with precipitation:")


Days with precipitation:


In [19]:
for obs in data["observations"]:
    if obs ["precipitation"] > 0 :
        print(f"Date: {obs['date']}, Precipitation: {obs['precipitation']}")
    
    

Date: 2023-01-02, Precipitation: 0.5
Date: 2023-01-04, Precipitation: 0.2


# Part 2

In [20]:
%pip install pooch

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [21]:
import pooch 

In [42]:
import os

In [43]:
file_path = pooch.retrieve(
    url = "https://github.com/pandas-dev/pandas/raw/main/doc/data/air_quality_no2.csv",
    known_hash=None)
print("File downloaded to:", file_path)
print ("File exists:", os.path.exists(file_path))

File downloaded to: /home/cam2463/.cache/pooch/458dad453f6a48e510cd544bef1854e3-air_quality_no2.csv
File exists: True


## Task 1

In [26]:
with open(file_path, "r") as f:
    for i in range(5):
        print(f.readline())

datetime,station_antwerp,station_paris,station_london

2019-05-07 02:00:00,,,23.0

2019-05-07 03:00:00,50.5,25.0,19.0

2019-05-07 04:00:00,45.0,27.7,19.0

2019-05-07 05:00:00,,50.4,16.0



In [29]:
file_size = os.path.getsize(file_path)
print(f"File size: {file_size} bytes")

File size: 31984 bytes


In [32]:
with open(file_path, "r") as f:
    line_count = len(f.readlines())

print(f" The file has {line_count} lines.")

 The file has 1036 lines.


## Task 2

In [49]:
with open(file_path, "r") as f:
    line_count =  len(f.readlines())

print(f"The file has {line_count} lines.")


The file has 1036 lines.


## Task 3

In [ ]:
print("Getting real weather data")
url = "https://api.weather.gov/points/40.7789,-73.9692"
response = requests.get(url)
if response.status_code == 200:
    weather_data = response.json()
    city = weather_data['properties'] ['relativeLocation'][ 'properties'][' city']
    state = weather_data['properties'][' relativeLocation' ]['properties' ]['state' ]
    print (f"Location: {city}, {state}")
    
    forecast_url = weather_data['properties']['forecast']
    forecast_response = requests.get(forecast_url)
    forecast = forecast_response.json()
   
    print("\nForecast:")
    print("-" * 50)
    for period in forecast["properties"]["periods"][:3]:
        print(f"{period['name']}:")
        print(f"Temperature: {period['temperature']}°{period["temperatureUnit"]}")
        print(f"Forecast: {period['shortForecast']}")
else:
    print(f"Error: {response.status_code}")

In [50]:
print("\nData Inventory:")
print("1. air_quality_no2.csv - Air quality NO2 measurements")


Data Inventory:
1. air_quality_no2.csv - Air quality NO2 measurements


# Part 3 

In [51]:
print("Getting dataset information... ")
base_url = print("Getting dataset information...")
base_url = "http://iridl.ldeo.columbia.edu/expert/SOURCES/.NOAA/.NCEP/.CPC/.UNIFIED_PRCP/.GAUGE_BASED/.GLOBAL/.v1p0/.Monthly/.RETRO/.rain/dods"
dds_url = base_url + ".dds"
response = requests.get(dds_url)

print("dataset Structure (DDS):")
print("=" * 60)
print(response.text[:500])
print("=" * 60)
dds_url = base_url +".dds"
response = requests.get(dds_url)

print("dataset Structure (DDS): ")
print("="*60)
print(response.text[:500])
print("="*60)


Getting dataset information... 
Getting dataset information...
dataset Structure (DDS):
Dataset {
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Float32 T[T = 324];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;

dataset Structure (DDS): 
Dataset {
    Float32 Y[Y = 360];
    Float32 X[X = 720];
    Float32 T[T = 324];
    Grid {
     ARRAY:
        Float32 rain[T = 324][Y = 360][X = 720];
     MAPS:
        Float32 T[T = 324];
        Float32 Y[Y = 360];
        Float32 X[X = 720];
    } rain;
} rain;



## Task 1

The dimensions are T 324, Y 360, x 720\
Rain is  the main variable

## Task 2

In [52]:
print("Getting dataset attributes...")
das_url = base_url + ".das"
response = requests.get(das_url)
print("Dataset Attributes (DAS):")
print("="*60)
print(response.text[:1000])
print("="*60)

Getting dataset attributes...
Dataset Attributes (DAS):
Attributes {
    Y {
        String standard_name "latitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 0;
        String units "degree_north";
    }
    X {
        String standard_name "longitude";
        Float32 pointwidth 0.5;
        Int32 gridtype 1;
        String units "degree_east";
    }
    T {
        Float32 pointwidth 1.0;
        String calendar "360";
        Int32 gridtype 0;
        String units "months since 1960-01-01";
    }
    rain {
        Int32 pointwidth 0;
        String standard_name "lwe_precipitation_rate";
        Float32 file_missing_value -999.0;
        String history "Boxes with less than 0.0% dropped";
        Float32 missing_value NaN;
        String units "mm/day";
        String long_name "Monthly Precipitation";
    }
NC_GLOBAL {
    String Conventions "IRIDL";
}
}



## Task 3

The dataset contains monthly rainfall measurements for the globe measured in millimeters per day. There is 27 years worth of data since January 1960. 